lib/lte_link_control/Kconfig

```diff
@@ -314,6 +334,10 @@ config LTE_NEIGHBOR_CELLS_MAX
 	  cells, so there's a trade-off between heap requirements and
 	  the risk of not being able to parse all neighbor cell information.
 
+endif # LTE_LC_NEIGHBOR_CELL_MEAS
+
+if LTE_LC_MODEM_SLEEP
+
 config LTE_LC_MODEM_SLEEP_NOTIFICATIONS
```

Comment:
So, I guess `LTE_LC_MODEM_SLEEP` enables the module, but `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` enables the feature?

But would anyone ever actually set `LTE_LC_MODEM_SLEEP=y` and `LTE_LC_MODEM_SLEEP_NOTIFICATIONS=n`?

Would it make more sense to simply use `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` directly to control whether the module is enabled?

I think this would be less confusing, too, since the module doesn't actually control whether modem sleep is enabled -- the module is specifically for processing modem sleep notifs

Reply:
> So, I guess `LTE_LC_MODEM_SLEEP` enables the module, but `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` enables the feature?

`LTE_LC_MODEM_SLEEP_NOTIFICATIONS` subscribes to the notifications.
`LTE_LC_MODEM_SLEEP` compiles and enables the module which can subscribe to the events and parse them but doesn't yet subscribe to the events.

> But would anyone ever actually set `LTE_LC_MODEM_SLEEP=y` and `LTE_LC_MODEM_SLEEP_NOTIFICATIONS=n`?

Modem shell does allow this and this is the default configuration because there is `link msleep -s | -u` command options for subscribing and unsubscribing the events. But it's a test application that doesn't simulate normal application too well in many aspects as we need to allow various things in there.

> Would it make more sense to simply use `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` directly to control whether the module is enabled?

It may be possible. We'll discuss about this tomorrow. This doesn't perfectly enable modem shell functionality but this would be good approach for all normal applications.

> I think this would be less confusing, too, since the module doesn't actually control whether modem sleep is enabled -- the module is specifically for processing modem sleep notifs

This is a valid point irrespective of whether `LTE_LC_MODEM_SLEEP` will remain, and must be clarified in the Kconfig comments and maybe in the Kconfig name.

Reply:
Aha, I see! So really, in essence, the difference between
```
LTE_LC_MODEM_SLEEP=y
LTE_LC_MODEM_SLEEP_NOTIFICATIONS=n 
```
and

```
LTE_LC_MODEM_SLEEP=y
LTE_LC_MODEM_SLEEP_NOTIFICATIONS=y 
```

is whether or not `LTE_LC` automatically subscribes to the notifications.

In which case, I suggest this instead:

Have `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` control whether the module is enabled, and then add a new KConfig `LTE_LC_MODEM_SLEEP_AUTO_SUBSCRIBE` to control whether `LTE_LC` automatically subscribes when initialized

Reply:
LTE Link Control has fundamental Kconfig option naming issues, starting from `CONFIG_LTE_LINK_CONTROL` to enable the library, and `CONFIG_LTE_*` for everything else. While it would be nice to have everything perfectly aligned, I'd prefer not to change the existing options, at least not in this PR.

Reply:
> LTE Link Control has fundamental Kconfig option naming issues, starting from `CONFIG_LTE_LINK_CONTROL` to enable the library and `CONFIG_LTE_LINK_CONTROL_LOG_LEVEL_*` to control logging, and `CONFIG_LTE_*` or `CONFIG_LTE_LC_*` for everything else. While it would be nice to have everything perfectly aligned, I'd prefer not to change the existing options, at least not in this PR.

There is a plan to homogenize the naming, but not for now.

Reply:
Yep, but the module names are something we should select now so that we don't need to rename them.
All other Kconfigs we shall not modify in this PR or in 2.8.0 context. This will make migration rather easy when only the required modules need to enabled by the clients.

Reply:
> This is problematic from the backwards compatibility and migration perspective.

Is that actually true? If you went with my suggestion , setting`LTE_LC_MODEM_SLEEP_NOTIFICATIONS=y` would still have the exact same effect as it always did.

Its a rename from the perspective of ***how the PR is currently written***

But the result would not really be a rename as compared to the state of NCS afterwards.



Here is another way to think of it that doesn't involve renaming:

1. Get rid of the newly introduced `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` setting
2. Make `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` control whether the module is enabled (this will arguably IMPROVE backward compatibility as compared to the current state of things)
3. Add a new setting: `LTE_LC_MODEM_SLEEP_REQ`, which defaults to `y`, which chooses whether the alerts are enabled at start

Reply:
> Just do this:
> 
> 1. Get rid of the newly introduced `LTE_LC_MODEM_SLEEP` setting
> 2. Make `LTE_LC_MODEM_SLEEP_NOTIFICATIONS` control whether the module is enabled
> 3. Add a new setting: `LTE_LC_MODEM_SLEEP_REQ`, which defaults to `y`, which chooses whether the alerts are enabled at start
> 
> No renaming needed. No backward compatibility issues.

Well, it's not a rename but you are changing the semantics of an existing Kconfig which cause issues somewhere.

Yep, I had `CONFIG_LTE_LC_MODEM_SLEEP_REQ`, like all other _REQ options, disabled by default in my mind but also figured out later that this would work quite nicely if this was 'y' by default (when `CONFIG_LTE_LC_MODEM_SLEEP_NOTIFICATIONS` are enabled). This handles `CONFIG_LTE_LC_MODEM_SLEEP_NOTIFICATIONS=y` case nicely.
However, when `CONFIG_LTE_LC_MODEM_SLEEP_NOTIFICATIONS=n` (as it is by default and in MOSH), run-time subscription to the events wouldn't work as before because the module is not compiled so you won't get lte_lc notifications. Not a major use case but theoretically possible someone else than MOSH does it. Good thing is that you'd get a compiler error due to the event enum flagging so you'll notice the change in behavior.

We are talking about renaming the module Kconfig's with `*_MODULE`. `CONFIG_LTE_LC_MODEM_SLEEP_MODULE` wouldn't anymore be quite as bad.

And another point why `CONFIG_LTE_LC_MODEM_SLEEP` is not completely "wrong" is because it maps to `AT%XMODEMSLEEP` which is also about notifications rather than enabling/disabling of the feature in modem side. So some/many people have concept of what it means if they've looked at AT documentation, and naturally we need to make it clear in lte_lc documentation too.
But we are going ahead with the `_MODULE` postfix.

Reply:
@glarsennordic Renamed to `*_MODULE` for better clarity.

Reply:
Fair enough about the semantic change. But yeah, module is a pretty good solution to my concerns I think, so I will resolve this